### PR TITLE
Add FlatPak recipe for Pioneer

### PR DESCRIPTION
**Pull request by pcercuei**
Pioneer is an open-source, multiplatform and fun space exploration game.

Signed-off-by: Paul Cercueil <paul@crapouillou.net>

### CODE REVIEW

File: net.pioneerspacesim.Pioneer.json [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	"sdk": "org.freedesktop.Sdk",
+	"command": "pioneer",
+
+	"finish-args": [
+		"--share=ipc",
+		"--socket=x11",
+		"--socket=wayland",
+		"--socket=pulseaudio",
+		"--device=dri",
+		"--persist=.pioneer"
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: Should be fixed upstream (and for flathub purposes, at least reported) to use $XDG_CONFIG_DIR.

File: net.pioneerspacesim.Pioneer.json [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	],
+
+	"build-options" : {
+		"cflags": "-O3",
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: Add `-g` too.

File: net.pioneerspacesim.Pioneer.json [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		"cxxflags": "-O3",
+		"strip": true,
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: Debug symbols are extracted into extension `appid.Debug` so makes no sense to disable it as it helps with debugging.

File: net.pioneerspacesim.Pioneer.json [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		"env": {
+			"PIONEER_DATA_DIR": "/app/usr/share/pioneer"
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: It's needed only during build or also for running?

File: net.pioneerspacesim.Pioneer.json [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+			"sources": [
+				{
+					"type": "archive",
+					"url": "https://github.com/assimp/assimp/archive/v4.1.0.tar.gz",
+					"sha256": "3520b1e9793b93a2ca3b797199e16f40d61762617e072f2d525fad70f9678a71"
+				}
+			]
+		},
+		{
+			"name": "libglu",
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: It's available in `shared-modules` repository that can be used as git submodule.

File: net.pioneerspacesim.Pioneer.json [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+					"url": "ftp://ftp.freedesktop.org/pub/mesa/glu/glu-9.0.0.tar.bz2",
+					"sha256": "1f7ad0d379a722fcbd303aa5650c6d7d5544fde83196b42a73d1193568a4df12"
+				}
+			]
+		},
+		{
+			"name": "pioneer",
+			"buildsystem": "autotools",
+			"build-commands": [
+				"install -Dm644 application-icon/pngs/pioneer-16x16.png /app/share/icons/hicolor/16x16/apps/net.pioneerspacesim.Pioneer.png",
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: You can use a regular bash loop here.

File: net.pioneerspacesim.Pioneer.json [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+				"install -Dm644 application-icon/pngs/pioneer-48x48.png /app/share/icons/hicolor/48x48/apps/net.pioneerspacesim.Pioneer.png",
+				"install -Dm644 application-icon/pngs/pioneer-64x64.png /app/share/icons/hicolor/64x64/apps/net.pioneerspacesim.Pioneer.png",
+				"install -Dm644 application-icon/pngs/pioneer-128x128.png /app/share/icons/hicolor/128x128/apps/net.pioneerspacesim.Pioneer.png",
+				"install -Dm644 application-icon/pngs/pioneer-256x256.png /app/share/icons/hicolor/256x256/apps/net.pioneerspacesim.Pioneer.png"
+			],
+			"sources": [
+				{
+					"type": "git",
+					"url": "https://github.com/pioneerspacesim/Pioneer.git",
+					"branch": "20180203"
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: We want builds to be reproducible so you need to specify a tag/commit here.

File: net.pioneerspacesim.Pioneer.json [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+		},
+		{
+			"name": "pioneer-metadata",
+			"buildsystem": "simple",
+			"build-commands": [
+				"install -Dm644 pioneer.desktop /app/share/applications/net.pioneerspacesim.Pioneer.desktop",
+				"install -Dm644 pioneer.appdata.xml /app/share/appdata/net.pioneerspacesim.Pioneer.appdata.xml"
+			],
+			"sources": [
+				{ "type": "file", "path": "pioneer.desktop" },
```
Comment (barthalion) [inline net.pioneerspacesim.Pioneer.json]: Both desktop and appdata files should be contributed upstream so other distributions can use them.

File: pioneer.appdata.xml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+			of it.
+		</p>
+		<p>Pioneer is under constant development and has a friendly
+			community of players, modders and developers around it.
+			We release a new version almost every day. We'd love
+			for you to try it and make part of the galaxy your own!
+		</p>
+	</description>
+
+	<categories>
```
Comment (barthalion) [inline pioneer.appdata.xml]: Categories are taken from the desktop file.

File: pioneer.desktop [hunk 10]
```diff
@@ -0,0 +1,10 @@
+[Desktop Entry]
+Version=1.0
+Name=Pioneer
+GenericName=Pioneer Space Simulator
+Comment=A game of lonely space adventure
+Exec=/app/bin/pioneer
```
Comment (barthalion) [inline pioneer.desktop]: `pioneer` should be enough but I can be wrong about it.

File: pioneer.appdata.xml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+			<image type="source">https://pioneerspacesim.net/assets/ss5.png</image>
+		</screenshot>
+		<screenshot>
+			<image type="source">https://pioneerspacesim.net/assets/ss7.png</image>
+		</screenshot>
+		<screenshot>
+			<image type="source">https://pioneerspacesim.net/assets/ss8.png</image>
+		</screenshot>
+	</screenshots>
+</component>
```
Comment (barthalion) [inline pioneer.appdata.xml]: It should also contain `releases` section.

File: net.pioneerspacesim.Pioneer.json (hunk 1 above)
Comment (pcercuei) [reply net.pioneerspacesim.Pioneer.json]: Well I don't have much control over that... And I doubt upstream will accept to change the savestates directory, that would confuse a lot of people.

File: net.pioneerspacesim.Pioneer.json (hunk 4 above)
Comment (pcercuei) [reply net.pioneerspacesim.Pioneer.json]: Only for building (then the path is saved within the executable).

File: net.pioneerspacesim.Pioneer.json (hunk 7 above)
Comment (pcercuei) [reply net.pioneerspacesim.Pioneer.json]: Actually it's a tag that corresponds to the 2018.02.03 release of the game.

File: net.pioneerspacesim.Pioneer.json (hunk 1 above)
Comment (barthalion) [reply net.pioneerspacesim.Pioneer.json]: It's not really rocket science though to read old saves from old directory, but use a correct one when there is no old save games. Please report it regardless.

File: net.pioneerspacesim.Pioneer.json (hunk 4 above)
Comment (barthalion) [reply net.pioneerspacesim.Pioneer.json]: :+1:

File: net.pioneerspacesim.Pioneer.json (hunk 7 above)
Comment (barthalion) [reply net.pioneerspacesim.Pioneer.json]: There's a `tag` keyword for this (I know it doesn't make a difference internally, but makes things less confusing.)

**Comment by impaktor** [comment]
Not sure, is there something I'm required to do here?

**Comment by barthalion** [comment]
Not really for now – in the future you will be able to push new release to [Flathub](https://flathub.org) by changing the tag in one json file. From things I requested above from the submitter is possibly making use of XDG spec to keep settings/save files in XDG_DATA_DIR/XDG_CONFIG_DIR and keeping appdata and desktop files in Pioneer git repo so other distributions can use them, but neither is a blocker if just issues or pull requests are filled.

File: net.pioneerspacesim.Pioneer.json (hunk 1 above)
Comment (pcercuei) [reply net.pioneerspacesim.Pioneer.json]: Reported here: https://github.com/pioneerspacesim/pioneer/issues/4392

**Comment by impaktor** [comment]
Changing to use XDG_DATA_DIR/XDG_CONFIG_DIR feels like a big intrusion, and defaulting to `~/.local/share` feels non-intuitive, as I would have guessed `~/.config/` is more common. 

Possibly the behavior of this could be set by compile flag at build time? That way whatever buld scripts you have could define where to put the save folder.

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/net.pioneerspacesim.Pioneer

**Comment by nedrichards** [comment]
Feel free to let us know about anyone else from upstream who should have the ability to push releases.

**Comment by impaktor** [comment]
> Feel free to let us know about anyone else from upstream who should have the ability to push releases.

Please feel free to also add @fluffyfreak, @laarmen, and @ecraven.

**Comment by nedrichards** [comment]
Done, thanks

**Comment by barthalion** [comment]
@nedrichards Unnecessary stripping and not used `shared-modules` were not solved. Let's not merge things that aren't clearly ready.

File: pioneer.appdata.xml (hunk 9 above)
Comment (pcercuei) [reply pioneer.appdata.xml]: @nedrichards, can I add it in the .desktop instead?
Like this:
```Categories=Game;AdventureGame;ActionGame;Simulation;```

Comment (TingPing) [reply pioneer.appdata.xml]: Yes that is where it belongs.

**Comment by impaktor** [comment]
@nedrichards could you also please add @johnbartholomew

**Pull request closed without merging**